### ISSUE TITLE
Results not being reproduced like original Neurogenesis paper

### ISSUE THREAD

**Issue by AmanGoyal99** [l-berg/bohb-over-growing-convnets#1]
Greetings,

So I have been using your repo from some time and after a lot of experiments trying to reproduce original paper as well, I am unable to get similar results.
Although test accuracy is also lesser, higher discrepancy is in the neuron generation trend. So in the original paper, neurons are generated steadily over various epochs.
However, using your paper reproducing code, neurons are generated within few iterations and maximum in first 5-10 epochs itself. 
So wanted to understand if something is missing in your implementation due to which neuron generation is so quick and even test accuracy is about 8-9% less for VGG.

Thanks,
Aman Goyal

**Comment by l-berg** [comment]
I assume you are 
1. calling the `GrowableModel` class with `reproduce_paper=True`, `max_params=None` and `trigger_threshold=0.9` in order to reproduce the NORTH-Select strategy. 
2. Then, if you want to use e.g. 2x VGG-11 as a maximum size (like in the paper), you set the `n_channels_conv_0=2*64`, `n_channels_conv_1=2*128` etc. (which is too large for my system).

I would expect the following: In the first layers, the number of neurons grows very fast, and only the later layers, where the maximum size is much larger, grow slower (see Figure 5a from the [original paper](https://openreview.net/attachment?id=SWOg-arIg9&name=main_paper_and_supplementary_material)).

One key difference in my implementation is that all layers have the same channel-count initially, whereas the original paper used the same proportions as the maximum size (meaning: `conv_0` starts from `64/2`, `conv_1` from `128/2`, etc.). 
Also, all my predefined blocks use max-pooling as a last step, therefore VGG-11 cannot be set as a base model without changing the `GrowableModel` class.

As you can read in my small [paper](https://github.com/l-berg/bohb-over-growing-convnets/blob/main/paper.pdf), I found some limitations of the NORTH-Select strategy. That is the reason why there is no exact reimplementation.

The first author of the original paper has recently set up a new repository which provides building blocks for the NORTH* strategies in PyTorch. Depending on what you are trying to do, it might be worth a look: https://github.com/SuReLI/NeurOps

If this does not answer your question, please provide further details on what parts of the code you are using and what arguments you set.

**Comment by AmanGoyal99** [comment]
Yes true, so the Growable class was slightly modified by me to have MaxPool layers like VGG-11. The problem is that not just the first, but many of the initial layers are growing too fast. They all reach max or stop within first 5 epochs. So my question will be why is the growth so fast?

**Comment by l-berg** [comment]
Have you tried starting from a scaled down version of VGG instead of all same-sized layers? Layers that have grown to a certain size differ from those that were randomly initialized at that size, because NORTH-Select adds neurons which maximize the effective rank metric, resulting in more growth in the future.
Also, maybe try using the same batch size as the original paper, as the metrics are computed across batches.

**Comment by AmanGoyal99** [comment]
Yes, so l had started from smaller version which is 1/4th of size of all layers and set 2 times as max channels. This was similar to what authors kept in the paper. 
When you run your version for reproducing paper, you will also be able to notice the instant growth.
This is my code: https://github.com/AmanGoyal99/Neurogenesis-Pruning-Pytorch/tree/develop/src

It is forked version of your repo. Just minor change is modification to have softmax layers in VGG-11. 
Would really appreciate if you could have a look once.
But I hope it is clear that the main issue here is the rapid generation of neurons. 
So that is quite visible from your original repo as well.

**Comment by l-berg** [comment]
Ok, after taking a look at your code and running it, I was unable to reproduce your problem. I trained for 50 epochs with batch-size 128 (which they used in the paper for VGG). Here are some of the results:
```
Epoch [1/50]  (=> Training conv[128-256-512-407-760-390-440-487]-fc[1024-1024] 1.53e+07) Loss: 2.0317
Epoch [5/50]  (=> Training conv[128-256-512-407-760-390-440-562]-fc[1024-1024] 1.59e+07) Loss: 0.9658
Epoch [50/50] (=> Training conv[128-256-512-407-1024-1024-1024-731]-fc[1024-1024] 3.68e+07) Loss: 0.0216
```

<img src="https://user-images.githubusercontent.com/89343112/221380765-0ac9d179-78d1-4292-9ad2-a31aee3d3f9e.png" width="50%"><img src="https://user-images.githubusercontent.com/89343112/221381296-2a82fbeb-e33c-4879-878b-050f4d823274.png" width="50%">

Maybe there was a misconception because the growth is very fast in the beginning and then only occurs sporadically later on?

**Comment by AmanGoyal99** [comment]
If you see other convolution layers, growth is pretty quick and maxes in first few iterations

**Comment by l-berg** [comment]
Please be more specific in which way this is different from the original paper. As you can see in figure 5, NORTH-Select maxes out the first layers immediately. None of the layers start from 0.25, because they all grow at the very start (see the blue line):
![image](https://user-images.githubusercontent.com/89343112/223473710-7e16548f-75c7-4d5c-9915-f174fc53da38.png)
In the end, most layers have grown to their maximum size after 100 epochs:
![image](https://user-images.githubusercontent.com/89343112/223474596-d1e2e402-43cd-455c-9d3b-849cd2c19ee0.png)

At the moment, I am not actively working in neurogenesis. Therefore, I do not have the resources to further prove or disprove whether my re-implementation is accurate without major hints for what might be wrong.

**Comment by AmanGoyal99** [comment]
What about other conv layers? Didn't you notice their quick growth within few iterations?

**Comment by AmanGoyal99** [comment]
I am attaching an image of the growth trends for all Conv layers obtained through the code which I had shared you earlier
![Slide1](https://user-images.githubusercontent.com/51092051/224241400-fd740f4e-67b3-4671-afcf-58f8afc2639d.jpg)

**Comment by AmanGoyal99** [comment]
Could you just copy and paste the hyperparameters you used when running the code I provided you? That would be helpful

**Comment by l-berg** [comment]
Your results do not appear to be that different from mine, only that you trained for WAY LESS iterations.

I used your [latest commit](https://github.com/AmanGoyal99/Neurogenesis-Pruning-Pytorch/tree/60f13b8706932e9b4207600da4ac54cbe662ea85) and this is the relevant output of git diff:

```diff
diff --git a/src/outer_bohb.py b/src/outer_bohb.py
index 73f3cc7..dcf0143 100644
--- a/src/outer_bohb.py
+++ b/src/outer_bohb.py
@@ -228,7 +228,8 @@ def evaluate_config(cfg: Configuration, seed: int, instance: str, budget: float,
     print('Total training time in hrs:',total_time/3600)
     ###Visualize neuron generation####
     # exit(0)
-    epochs = [i+1 for i in range(len(train_loader)*10)]
+    #epochs = [i+1 for i in range(len(train_loader)*10)]
+    epochs = [i+1 for i in range(len(edim_track[0]))]
     visualize(epochs,track_growth[0],'Conv1_neuron_growth','iterations','Relative Layer Sizes')
     visualize(epochs,track_growth[1],'Conv2_neuron_growth','iterations', 'Relative Layer Sizes')
     visualize(epochs,track_growth[2],'Conv3_neuron_growth','iterations', 'Relative Layer Sizes')
@@ -379,15 +380,15 @@ if __name__ == '__main__':
         'reproduce_paper': True,
         'learning_rate_init': 3e-4,
         # 'learning_rate_init': 0.0001,
-        'batch_size': 256,
+        'batch_size': 128,
         # 'batch_size': 100,
         'data_dir': 'CIFAR10',
         'optimizer': 'Adam',
         'train_criterion': 'CrossEntropy',
-        'device':'cuda'
+        'device':'cpu'
     }
     cfg = default_config
-    evaluate_config(cfg, 0, '', budget=2, nfolds=2)
+    evaluate_config(cfg, 0, '', budget=50, nfolds=2)
     # # SMAC scenario object
     # scenario = Scenario({"run_obj": "quality",  # optimize quality (alternative to runtime)
     #                      "wallclock-limit": runtime,  # max duration to run the optimization (in seconds)
```

**Comment by AmanGoyal99** [comment]
conv[128-256-512-389-713-386-519-860]-fc[2078-1968] 2.53e+07) Loss: 0.0291

This what I get after 50 epochs and batchsize as 128. Did you run using my Growablemodels file ? Also I am getting test acc of 69%. What about you?

**Comment by AmanGoyal99** [comment]
What test accuracy are you getting ?

**Comment by l-berg** [comment]
Does your code record the test accuracy by default? If not, I am afraid I cannot tell you. I just ran your code with the modifications mentioned above (`python src/outer_bohb.py --device=cpu`). This took an entire day on my laptop CPU.

I did not verify whether my implementation can reproduce the numbers from the paper because of the high computational costs. Now, I simply do not have the time nor compute resources to look further into this. At this point, I can only refer you to the author's [reference implementation](https://github.com/k8lion/Neurogenesis).

**Comment by AmanGoyal99** [comment]
Yes yes, it records. Just scroll at the bottom of your results. There is test accuracy calculated.

**Comment by AmanGoyal99** [comment]
Have you used the same growable models.py file which I have in my repo
